chalk-recursive/src/coinduction_handler.rs

```diff
@@ -0,0 +1,197 @@
+use ::std::cmp::{max, min};
+use chalk_ir::{interner::Interner, Canonical, ConstrainedSubst, Constraints, Fallible};
+use chalk_solve::Solution;
+use rustc_hash::FxHashMap;
+use std::collections::BinaryHeap;
+use tracing::debug;
+
+use crate::{
+    search_graph::{DepthFirstNumber, SearchGraph},
+    Minimums, UCanonicalGoal,
+};
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct CoinductiveCycleDependencyBoundaries {
+    pub(crate) lower: DepthFirstNumber,
+    pub(crate) upper: DepthFirstNumber,
+}
+
+impl CoinductiveCycleDependencyBoundaries {
+    pub(crate) fn update_from(&mut self, other_boundaries: CoinductiveCycleDependencyBoundaries) {
+        self.lower = min(self.lower, other_boundaries.lower);
+        self.upper = max(self.upper, other_boundaries.upper);
+    }
+
+    pub(crate) fn eq_singular(&self, singular: DepthFirstNumber) -> bool {
+        self.lower == singular && self.upper == singular
+    }
+
+    pub(crate) fn new(singular: DepthFirstNumber) -> Self {
+        CoinductiveCycleDependencyBoundaries {
+            lower: singular,
+            upper: singular,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct PrematureResult<I: Interner> {
+    result: Fallible<Solution<I>>,
+
+    /// The outermost and innermost cycles on which this result depends.
+    dependency: CoinductiveCycleDependencyBoundaries,
+}
+
+pub(crate) struct CoinductionHandler<I: Interner> {
+    /// Heap of start DFNs for nested cycles.
+    /// The innermost cycle start is the first element in the heap.
+    cycle_start_dfns: BinaryHeap<DepthFirstNumber>,
```

Comment:
This heap could also be vector that then needs to be sorted after every insertion. Otherwise, the order in which goals are evaluated would matter for intercycle dependencies.